### PR TITLE
Prototype component remover

### PR DESCRIPTION
**Pull request by Krunklehorn**
## About the PR

Adds a new system to remove other components in response to an event.
Currently a draft to help workshop additional features or concerns.

## Why / Balance

I was adding RemComp lines to prevent the Exterminator antagonist from getting hungry or thirsty when Sloth suggested the pattern was common enough to be turned into a component.

## Technical details

Example:

Instead of...

```
private void OnMapInit(EntityUid uid, SomeComponent comp, MapInitEvent ev)
{
    RemComp<HungerComponent>(uid);
    RemComp<ThirstComponent>(uid);
}
```

...you can do...

```
- type: entity
  abstract: true
  id: MobTerminatorBase
  components:
  - type: ZombieImmune # yeah no
  - type: FlashImmunity # no brain to brainwash, eyes are robotic
  - type: ComponentRemover
    doOnMapInit: true
    components:
    - type: Hunger
    - type: Thirst
```

Currently only supports MapInitEvent. -> Could this be made generic? 🤔
If not, please suggest other events this pattern is commonly found in.

Fails silently for any listed component that doesn't exist. -> Is this desired?

## Media

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

### CODE REVIEW

**Comment by deltanedas** [comment]
> prevent the Exterminator antagonist from getting hungry or thirst

i already have this implemented in a way that they can still eat in other pr

**Comment by deltanedas** [comment]
also i think this should just be a prototype thing like for mapping since a ***lot*** of things only exist to not have a component that would otherwise be inherited (syndie monkey and base monkey for example)

**Comment by Krunklehorn** [comment]
Some prototypes may be easier to approach from a subtractive angle, rather than additive.

MobMonkey vs. MobBaseSyndicateMonkey is 3 new components + 7 copied components. It would be shorter to define as MobMonkey + NpcFactionMember + Loadout - Clumsy.
I wouldn't be surprised if their are more egregious examples.

**Comment by Krunklehorn** [comment]
After taking a look myself, I didn't actually find many existing _systems_ that would use this.

Most _systems_ that use RemComp call it after some condition checks. It's rarely done to avoid mass YAML duplication, like TerminatorSystem.

Specifically for OnMapInit, AutoImplantSystem and SpawnRandomOffsetSystem both perform additional behavior.

5 other systems call RemComp plainly, but these should remain code since they occur midway during the entity's lifetime...
ReagentGrinderSystem, PathfindingSystem, VoiceMaskSystem.Equip, MicrowaveSystem, SharedJitteringSystem



Automatic component removal really only shines with helping cleanup heavy YAML duplication, but as delta suggested, if it's only for OnMapInit, it should be a part of an existing system.


## TL;DR
Should automatic component removal be handled by the prototype serializer or the entity spawner?

**Comment by ElectroJr** [comment]
If possible, prototypes should be arranged so that "subtraction" shouldn't be required, which can often just be done by using abstract parents or composition / multiple parents. Being able to just remove components might encourage badly assembled prototypes/inheritance and lead to more spaghetti.  

I don't really feel strongly about this tough, and wouldn't mind being able to remove prototypes, but if it does get added it should be done when entity prototypes are loaded using a datafield on `EntityPrototype`, not when entities get spawned using a component & event.

> Some prototypes may be easier to approach from a subtractive angle, rather than additive.
> 
> MobBaseSyndicateMonkey is 3 new components + 7 copied components. It would be shorter to define as MobMonkey + NpcFactionMember + Loadout - Clumsy.

That example could just as well be handled by introducing an abstract `BaseMobMonkey` and then using
`MobMonkey = BaseMobMonkey + Clumsy` and `MobBaseSyndicateMonkey = BaseMobMonkey + Loadout + NpcFactionMember`.

**Comment by deltanedas** [comment]
> That example could just as well be handled by introducing an abstract BaseMobMonkey and then using

which is what is done, but it would be more obvious with removing the clumsy component from the syndie prototype

**Comment by metalgearsloth** [comment]
Agree with Electro if this does go anywhere.

To my knowledge the only place that it's really done atm is CollisionWake being toggled off atm, anything else can be done additively.
